First Draft Implementation of Oras Python!

vsoch opened the pull request:
Hey everyone! I've finished up the "what I think are hard requirements" from #5 and here is a summary of the first round of work that will provide a working oras client. Some notes:

 - Basic support for login/logout/push/pull with and without authentication
 - Artifacts can be single blobs or directories, the latter of which get a different media type and are tar-guzzed up and extracted on download. We get the filename from an annotation.
 - Basic tests for linting and core functionality is added, with the exception of authenticated views. I don't have an easy way to set up a full auth server in an action, so the most we could do right now is to add a workflow that brings up a server with a password and then we test login/logout. I'm happy to do that in another PR (after this one) and if it's desired give me a +1 and I'll open an issue.
 - Functions have docstrings and parameters defined, along with typing
 - We lint-all-the-things, including black, pyflakes, mypy, and isort. 
 - A Makefile makes install / lint easy for the developer
- There is a release workflow added already with a credential to pypi, so making a release of the software here releases to pypi.
- I've made pretty branded docs that match oras.land (and they link there) and I'd like to extend this to include Python docstrings, just want to play around with some ideas before just defaulting to standard sphinx (will open issue after merge here). Oh! And would be super awesome to get the docs previews hooked up to Netlify? I love Netlify but none of my academic projects use it, because the free tier is too small and we cheap. :laughing: 
- Testing is for functionalty and linting, of course
- Manifests and other json dumps are validated with jsonschema
- An automated container build to deploy to ghcr.io/oras-project/oras-py:latest (or a tag on release)

And there are still a list of things to discuss under #5 - e.g., some details that aren't yet added from oras-go (e.g., envars to drive things) that I think would be best if someone familiar could go through and add a list to an issue, and I'll tackle more slowly over time. I do think it's important to get something up sooner than later so the community can give feedback and contribute!

With respect to "copy" - I know the original oras didn't have this design and it was refactored, but I'm not convinced yet that this abstraction is useful. That said, if someone comes to me and says "I need this to be a more abstract storage that I can copy to" we can refactor to generalize. My spidey sense is that 90% of people just want to push/pull from standard registries, and so I followed the logic of "keep it simple, stupid" and implemented that.

After this original round of work I'm going to be working with @wolfv to discuss his needs to use oras with Mamba/Conda, and make sure the library here meets them. I suspect we will be putting together the first "custom client example" once we have that chat (and we can open a separate issue for this too).

Yay! :partying_face: Sorry I should have done this from the getgo, I wasted a lot of time trying a really bad idea, sort of assumed someone else would step up, and realized two days ago nobody was so I'd just hammer it out. :hammer:

sajayantony commented:
Now we need python reviewers :) But I'm in favor or merging into main first and incrementally addressing things through issues or smaller PRs.

vsoch commented:
I'm good with that too @sajayantony ! I am here at your service for all the issues / feature requests that come in, and I'm going to keep working on the things I noted above, and probably think of more as I go. I'm not experienced with oras-go but it would be fairly easy to look there and get a sense of what I'm missing, environment variable and interaction wise. I also want to get that GitHub example up with developer docs (for using the Python API) sooner than later.

Happy Thursday!

sajayantony commented:
@jdolitsky @deitch

pbarker commented on oras/client.py:
```diff
@@ -0,0 +1,139 @@
+
+    def __init__(
+        self,
+        hostname: str = None,
```
Optionals?

pbarker commented on oras/logger.py:
```diff
@@ -0,0 +1,313 @@
+        isatty = getattr(self.stream, "isatty", None)
+        return isatty and isatty()  # type: ignore
+
+    def emit(self, record):
```
type?

pbarker commented on oras/logger.py:
```diff
@@ -0,0 +1,313 @@
+        for handler in self.log_handler:
+            handler(msg)
+
+    def set_stream_handler(self, stream_handler):
```
type?

pbarker commented on oras/schemas.py:
```diff
@@ -0,0 +1,65 @@
+
+# Layer
+
+layerProperties = {
```
may be fun to generate these in the future

pbarker commented on oras/tests/artifact.txt:
```diff
@@ -0,0 +1 @@
+hello dinosaur
```
🦕

pbarker commented on oras/utils/fileio.py:
```diff
@@ -0,0 +1,262 @@
+    return dest_name
+
+
+def extract_targz(targz, outdir):
```
type

pbarker reviewed (commented):
Beautiful code! Thanks for doing this!

vsoch commented on Dockerfile:
```diff
@@ -0,0 +1,17 @@
+FROM ubuntu:22.04
+
+# docker build -t oras-py .
```
Yes I always comment my Dockerfile with the build command for the developer user to quickly see the intended build.

vsoch commented on oras/auth.py:
```diff
@@ -0,0 +1,87 @@
+from oras.logger import logger
+
+
+def load_configs(configs: List[str] = None):
```
In theory yes, although we always provide from the client so it always comes as a list. But I'm happy to tweak this to be Optional so it's more correct!

vsoch commented:
@pbarker thank you for the review! I've fixed everything locally and will likely do a push with changes after I worked a bit more on the requests in #11 (which has some great ideas that I think will make it better!)

sajayantony reviewed (approved):
LTGM to merge and iterate on this going forward.

vsoch commented:
Thank you @sajayantony ! Let's hold off merging until I get the final tweaks in this evening, if that's ok :)

sajayantony commented:
@vsoch - If there is one thing I have learned, it is "don't merge too fast"  :) I've been burnt a few times.

vsoch commented:
I'm going to wait for this to be merged until I start playing around with a documentation strategy - this PR is getting too chonky (well it's already superchonk) and I'm mostly worried about messing something up.

wolfv commented:
This looks great! I would be happy to help maintain the `oras-py` client going forward. Just to give some context, we're looking into using the Github OCI registry as mirror / backup and we want to offer "native" support in the mamba package manager asap.

vsoch commented:
yay! That sounds super fun! :partying_face: I'm happy to help maintain too, and continue tweaking for your use case (I'm a huge fan of mamba/conda btw!) 

And I just added spell checking, contributor stats, and some super pretty docs - they match the oras.land branding but of course have more detailed usage (and docstrings) for the Python API here. I also updated blob -> blob_path, not the perfect solution to your issue but works for the time being.

![image](https://user-images.githubusercontent.com/814322/169664203-52f81070-b569-4ba0-84ab-115b9182e1ba.png)

And here is what the rendered docstrings look like. The "sphinx format" isn't as pretty to see in the text imho (and it's redundant to have the type again) but it's a solution for now until/unless we decide on something better. I do think a Python API should have the basic usage / source readily available so people don't have to search through GitHub.

![image](https://user-images.githubusercontent.com/814322/169664237-502f1968-603b-4a1a-8b57-29e163daf4a4.png)

wolfv commented on Dockerfile:
```diff
@@ -0,0 +1,17 @@
+ENV PATH /opt/conda/bin:${PATH}
+ENV LANG C.UTF-8
+RUN apt-get update && \
+    apt-get install -y wget && \
```
coming from the mamba world I'd advocate for the `micromamba` docker image here, and using the `python` packages from the `conda-forge` channel: https://github.com/mamba-org/micromamba-docker/ – you could also install `ipython` from conda-forge :)

vsoch replied:
Haha, fair! We could probably add another recipe for that if you think it’s much better, even elsewhere. My preference generally is for more vanilla base images but you are right I broke that rule a bit with conda instead of installing python3. We can have a discussion with other maintainers to gauge preferences - I’d be ok with any of these ideas really!

wolfv replied:
I guess if you want to use `pip` anyways, we could also use the official `python` base images? (https://hub.docker.com/_/python)

wolfv commented:
I think this looks great. Should it be merged?

I would be happy to try it out in our OCI mirroring actions (we're currently calling `oras` in a python subprocess...). This will be much more convenient.

sajayantony reviewed (approved):
Initial draft for ORAS-py

vsoch commented:
Woot! If I'm allowed to merge my own PRs (at least for this draft) let's do it. I'll follow up quickly if there is any issue with the docs build, etc.
